### PR TITLE
Update status classes to allow additional levels of aggregation and to allow hiding subcomponents

### PR DESCRIPTION
**Pull request by scottbommarito**
https://github.com/NuGet/NuGetGallery/issues/6377
https://github.com/NuGet/NuGetGallery/issues/6378

Previously:
- each incident is linked to an event
- each event is associated with an component
- each message is linked to an event

This has become noisy, because we do not want to emit messages for every component (e.g. we shouldn't report for gallery USSC).

With this change:
- each incident is linked to an incident group
- each incident group is associated with a component
- each incident group is linked to an event
- each event is associated with a component
- each message is linked to an event

This will allow us to control what components for which messages are posted, and to post messages that describe the downtime of multiple components rather than just single components.

### CODE REVIEW

File: src/NuGet.Services.Status.Table/IChildEntity.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+namespace NuGet.Services.Status.Table
+{
+    /// <summary>
+    /// An entity that has a parent.
+    /// </summary>
+    public interface IChildEntity<T>
+    {
+        string ParentRowKey { get; }
+
+        bool IsLinked { get; }
```
Comment (joelverhagen) [inline src/NuGet.Services.Status.Table/IChildEntity.cs]: xml docs

File: src/NuGet.Services.Status.Table/IChildEntity.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace NuGet.Services.Status.Table
+{
+    /// <summary>
+    /// An entity that has a parent.
+    /// </summary>
+    public interface IChildEntity<T>
```
[the remark this answers is not in the document]
Comment (chenriksson) [reply src/NuGet.Services.Status.Table/IChildEntity.cs]: Also, trying to determine if we *really* use this interface anywhere... maybe worth reviewing?

Looks like there are 2 `IChildEntity` types: `ChildEntity` and `AggregatedEntity`. `AggregatedEntity` uses `ChildEntity` but never `IChildEntity`.

File: src/NuGet.Services.Status.Table/AggregatedEntity.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace NuGet.Services.Status.Table
+{
+    /// <summary>
+    /// Base implementation of <see cref="IAggregatedEntity{T}"/>.
+    /// </summary>
+    public class AggregatedEntity<T> : ComponentAffectingEntity, IAggregatedEntity<T>
```
[the remark this answers is not in the document]
Comment (scottbommarito) [reply src/NuGet.Services.Status.Table/AggregatedEntity.cs]: I like `AggregatedComponentAffectingEntity`

File: src/NuGet.Services.Status.Table/IChildEntity.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (scottbommarito) [reply src/NuGet.Services.Status.Table/IChildEntity.cs]: The generic arg allows strong typing requirements on the parent entity.

All the interfaces in this PR are used in the status job PR.

File: src/NuGet.Services.Status.Table/IncidentGroupEntity.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+{
+    /// <summary>
+    /// An aggregation of all of the <see cref="IncidentEntity"/>s affecting a single component during a time period.
+    /// Is aggregated by <see cref="EventEntity"/>.
+    /// </summary>
+    public class IncidentGroupEntity : AggregatedEntity<EventEntity>
+    {
+        public const string DefaultPartitionKey = "groups";
+
+        public IncidentGroupEntity()
```
[the remark this answers is not in the document]
Comment (scottbommarito) [reply src/NuGet.Services.Status.Table/IncidentGroupEntity.cs]: We can't really provide a guarantee that the `RowKey` and `PartitionKey` of a `TableEntity` doesn't change--

File: src/NuGet.Services.Status/Component.cs [hunk 5]
```diff
@@ -14,27 +14,36 @@ namespace NuGet.Services.Status
     /// An <see cref="IComponent"/> should calculate its <see cref="Status"/> based on its <see cref="SubComponents"/>.
     /// Different subclasses of <see cref="Component"/> should calculate their <see cref="Status"/> differently.
     /// </remarks>
-    public abstract class Component : ReadOnlyComponent, IComponent
+    public abstract class Component : IComponent
     {
-        public new abstract ComponentStatus Status { get; set; }
-        public new IEnumerable<IComponent> SubComponents { get; }
+        public string Name { get; }
+        public string Description { get; }
```
[the remark this answers is not in the document]
Comment (scottbommarito) [reply src/NuGet.Services.Status/Component.cs]: On the status page, as a description of the component.

image

"Downloading and installing packages from NuGet" and "Restore using the V3 API" are the descriptions of the two components.

File: src/NuGet.Services.Status/Constants.cs [hunk 6]
```diff
@@ -6,7 +6,7 @@ namespace NuGet.Services.Status
     public static class Constants
```
[the remark this answers is not in the document]
Comment (scottbommarito) [reply src/NuGet.Services.Status/Constants.cs]: I looked at the status job, and it doesn't use it. Making internal.

[the remark this answers is not in the document]
Comment (scottbommarito) [reply src/NuGet.Services.Status/Constants.cs]: Actually, `NuGet.Services.Status` and `NuGet.Services.Status.Table` are separate projects, so either I have to do the "internals visible to" shenanigans or I leave this as `public`.

File: src/NuGet.Services.Status.Table/IComponentAffectingEntity.cs [hunk 7]
```diff
@@ -0,0 +1,27 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace NuGet.Services.Status.Table
+{
+    /// <summary>
+    /// An entity that describes a period of time during which a component was affected.
```
[the remark this answers is not in the document]
Comment (scottbommarito) [reply src/NuGet.Services.Status.Table/IComponentAffectingEntity.cs]: I added comments for each property so it should be self-explanatory now.

File: src/NuGet.Services.Status.Table/IAggregatedEntity.cs [hunk 8]
```diff
@@ -0,0 +1,13 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace NuGet.Services.Status.Table
+{
+    /// <summary>
+    /// A <see cref="IComponentAffectingEntity"/> that is aggregated by <typeparamref name="T"/>.
+    /// </summary>
+    public interface IAggregatedEntity<T> : IChildEntity<T>, IComponentAffectingEntity
```
[the remark this answers is not in the document]
Comment (scottbommarito) [reply src/NuGet.Services.Status.Table/IAggregatedEntity.cs]: Using @chenriksson's suggestion of `IAggregatedComponentAffectingEntity`.

File: src/NuGet.Services.Status.Table/IChildEntity.cs (hunk 1 above)
Comment (scottbommarito) [reply src/NuGet.Services.Status.Table/IChildEntity.cs]: done

File: src/NuGet.Services.Status.Table/AggregatedEntity.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (scottbommarito) [reply src/NuGet.Services.Status.Table/AggregatedEntity.cs]: using `TAggregation`

File: src/NuGet.Services.Status.Table/AggregatedEntity.cs [hunk 9]
```diff
@@ -0,0 +1,52 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace NuGet.Services.Status.Table
+{
+    /// <summary>
+    /// Base implementation of <see cref="IAggregatedEntity{T}"/>.
```
[the remark this answers is not in the document]
Comment (scottbommarito) [reply src/NuGet.Services.Status.Table/AggregatedEntity.cs]: I updated the comments and did my best to make them more descriptive.

File: src/NuGet.Services.Status.Table/ChildEntity.cs [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.WindowsAzure.Storage.Table;
+
+namespace NuGet.Services.Status.Table
+{
+    /// <summary>
+    /// Base implementation of <see cref="IChildEntity{T}"/>.
+    /// </summary>
+    public class ChildEntity<T> : TableEntity, IChildEntity<T>
```
[the remark this answers is not in the document]
Comment (scottbommarito) [reply src/NuGet.Services.Status.Table/ChildEntity.cs]: using `TParent`

File: src/NuGet.Services.Status/ComponentUtility.cs [hunk 11]
```diff
@@ -82,5 +97,114 @@ public GetByNamesHelper(TComponent root)
...
+            foreach (var subcomponent in component.SubComponents.SelectMany(s => s.GetAllComponents()))
+            {
+                yield return subcomponent;
+            }
+        }
+
+        /// <summary>
+        /// Returns the deepest common ancestor of <paramref name="firstComponent"/> and <paramref name="secondComponent"/>.
+        /// </summary>
+        public static string GetLeastCommonAncestorPath<TComponent>(TComponent firstComponent, TComponent secondComponent)
```
[the remark this answers is not in the document]
Comment (scottbommarito) [reply src/NuGet.Services.Status/ComponentUtility.cs]: It's used in the jobs PR in the messaging logic.

File: src/NuGet.Services.Status/ComponentUtility.cs [hunk 12]
```diff
@@ -82,5 +97,114 @@ public GetByNamesHelper(TComponent root)
...
+        public static IEnumerable<IComponent> GetAllComponents(this IComponent component)
+        {
+            if (component == null)
+            {
+                yield break;
+            }
+
+            yield return component;
+
+            if (!component.DisplaySubComponents)
```
[the remark this answers is not in the document]
Comment (scottbommarito) [reply src/NuGet.Services.Status/ComponentUtility.cs]: `DisplaySubComponents` determines whether certain components are included in the `IReadOnlyComponent` built from an `IComponent`.

File: src/NuGet.Services.Status.Table/ComponentAffectingEntity.cs [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+            AffectedComponentStatus = (int)affectedComponentStatus;
+            StartTime = startTime;
+            EndTime = endTime;
+        }
+
+        public string AffectedComponentPath { get; set; }
+
+        /// <remarks>
+        /// This should be a <see cref="ComponentStatus"/> converted to an enum.
+        /// See https://github.com/Azure/azure-storage-net/issues/383
```
[the remark this answers is not in the document]
Comment (scottbommarito) [reply src/NuGet.Services.Status.Table/ComponentAffectingEntity.cs]: To be honest, I worry that messing around with the `Flatten` and `ConvertBack` methods would be more trouble than it's worth--dealing with enum conversion was fairly easy.

[the remark this answers is not in the document]
Comment (scottbommarito) [reply src/NuGet.Services.Status.Table/ComponentAffectingEntity.cs]: Furthermore, I think making multiple properties (enum and int) could become fairly confusing, I think it's clearer this way.

File: src/NuGet.Services.Status/ComponentUtility.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (scottbommarito) [reply src/NuGet.Services.Status/ComponentUtility.cs]: I looked at this again, it seems this method returns only VISIBLE subcomponents...I will improve the name

**Pull request merged**